From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manikanta Maddireddy <mmaddireddy@nvidia.com>
Date: Thu, 12 Jan 2023 11:17:24 +0530
Subject: [PATCH 222/245] NVIDIA: SAUCE: PCI: tegra194: Disable interrupt
 before link status check

Data abort is observed due to synchronization issue between irq
handler and driver shutdown call. Disable interrupt before link
status check to avoid this issue.

bug 3938884

Change-Id: I95b071eac5eb3317a4218920834c75e77289150f
Signed-off-by: Manikanta Maddireddy <mmaddireddy@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-5.10/+/2850134
(cherry picked from commit 32c76dae3de74c3f7902049d51c74e3d0aed9d7e)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/2905660
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2964976
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Reviewed-by: Prathamesh Shete <pshete@nvidia.com>
Tested-by: Prathamesh Shete <pshete@nvidia.com>

diff --git a/drivers/pci/controller/dwc/pcie-tegra194.c b/drivers/pci/controller/dwc/pcie-tegra194.c
index 304c0a9796b21..02fd890b3ac02 100644
--- a/drivers/pci/controller/dwc/pcie-tegra194.c
+++ b/drivers/pci/controller/dwc/pcie-tegra194.c
@@ -2480,6 +2480,7 @@ static void tegra_pcie_dw_remove(struct platform_device *pdev)
 	struct dw_pcie_ep *ep = &pcie->pci.ep;
 
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE) {
+		disable_irq(pcie->prsnt_irq);
 		if (!pcie->link_state)
 			return;
 
@@ -2622,17 +2623,15 @@ static void tegra_pcie_dw_shutdown(struct platform_device *pdev)
 	struct tegra_pcie_dw *pcie = platform_get_drvdata(pdev);
 
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE) {
-		if (!pcie->link_state)
-			return;
-		if (!pm_runtime_enabled(pcie->dev))
-			return;
-
 		debugfs_remove_recursive(pcie->debugfs);
 		disable_irq(pcie->prsnt_irq);
 		disable_irq(pcie->pci.pp.irq);
 		if (IS_ENABLED(CONFIG_PCI_MSI))
 			disable_irq(pcie->pci.pp.msi_irq[0]);
-
+		if (!pcie->link_state)
+			return;
+		if (!pm_runtime_enabled(pcie->dev))
+			return;
 		tegra_pcie_dw_pme_turnoff(pcie);
 		tegra_pcie_unconfig_controller(pcie);
 		pm_runtime_put_sync(pcie->dev);
-- 
2.43.0

